File: lintcode/binary-tree-preorder-traversal.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/binary-tree-preorder-traversal/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None

class Solution:
    def work_with_recursion(self, root, result):
        if not root:
            return
        result.append(root.val)
        self.work_with_recursion(root.left, result)
        self.work_with_recursion(root.right, result)

    def work_without_recursion(self, root):
        result = []
        if not root:
            return result

        stack = []
        stack.append(root)
        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        return result

    """
    @param root: The root of binary tree.
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        # write your code here
        return self.work_without_recursion(root)
        # result = []
        # self.work_with_recursion(root, result)
        # return result

if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    s = Solution()
    print s.preorderTraversal(root)

File: leetcode/binary-tree-preorder-traversal.cc
// Question: https://leetcode.com/problems/binary-tree-preorder-traversal/

#include "headers.h"

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        return iterate(root);

        // vector<int> ans;
        // recur(root, ans);
        // return ans;
    }

private:
    void recur(TreeNode* root, vector<int>& ans) {
        if (!root) {
            return;
        }
        ans.push_back(root->val);
        recur(root->left, ans);
        recur(root->right, ans);
    }

    vector<int> iterate(TreeNode* root) {
        vector<int> ans;
        if (!root) {
            return ans;
        }

        stack<TreeNode*> s;
        s.push(root);
        while (!s.empty()) {
            TreeNode* node = s.top();
            s.pop();
            ans.push_back(node->val);
            if (node->right) {
                s.push(node->right);
            }
            if (node->left) {
                s.push(node->left);
            }
        }

        return ans;
    }
};

int main(int argc, char** argv) {
    Solution s;
    {
        TreeNode* root = new TreeNode(100);
        root->left = new TreeNode(200);
        root->left->left = new TreeNode(300);
        root->left->right = new TreeNode(1000);
        root->right = new TreeNode(400);
        root->right->left = new TreeNode(500);
        vector<int> ans = s.preorderTraversal(root);
        for (int i : ans) {
            cout << i << " ";
        }
        cout << endl;
    }
}
